File: ios/MyCommute/WidgetModule.m
#import <TargetConditionals.h>

#if TARGET_OS_IPHONE

#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#else
#import "../../node_modules/react-native/React/Base/RCTBridgeModule.h"
#endif

@interface RCT_EXTERN_MODULE(WidgetModule, NSObject)
  RCT_EXTERN_METHOD(reloadWidget:(NSString *)jsonString
                    resolver:(RCTPromiseResolveBlock)resolve
                    rejecter:(RCTPromiseRejectBlock)reject)
  RCT_EXTERN_METHOD(saveWidgetStatusCache:(NSString *)jsonString
                    resolver:(RCTPromiseResolveBlock)resolve
                    rejecter:(RCTPromiseRejectBlock)reject)
@end

#endif

File: ios/MyCommute/WidgetModule.swift
import Foundation
import WidgetKit
import React

@objc(WidgetModule)
class WidgetModule: NSObject {
  
  @objc
  func reloadWidget(_ jsonString: String,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let userDefaults = UserDefaults(suiteName: "group.com.mycommute.app") else {
      reject("WIDGET_ERROR", "Failed to access shared UserDefaults group (group.com.mycommute.app)", nil)
      return
    }
    
    userDefaults.set(jsonString, forKey: "myLines")
    
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadAllTimelines()
      print("⚡️ Widget Kicked: Reloading Timelines")
    }
    resolve(nil)
  }
  
  @objc
  func saveWidgetStatusCache(_ jsonString: String,
                             resolver resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let userDefaults = UserDefaults(suiteName: "group.com.mycommute.app") else {
      reject("WIDGET_ERROR", "Failed to access shared UserDefaults group (group.com.mycommute.app)", nil)
      return
    }
    
    userDefaults.set(jsonString, forKey: "cachedTfLStatus")
    
    if #available(iOS 14.0, *) {
      WidgetCenter.shared.reloadAllTimelines()
      print("⚡️ Widget Status Cached: Reloading Timelines")
    }
    resolve(nil)
  }
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
}
